Encryption.md

```diff
@@ -133,48 +159,144 @@ Algorithms section. The length of the encryption buffer (a 4-byte little endian)
 with the same key. In other cases - when column(s) and the footer are encrypted with 
 different keys; or column(s) are encrypted and the footer is not - an extra measure is 
 required to protect the column-specific information in the file footer. In these cases, 
 the column-specific information (kept in `ColumnMetaData` structures) is split from the 
-footer, by utilizing the `required i64 file_offset` parameter in the `ColumnChunk` 
-structure. This allows to serialize each `ColumnMetaData` structure separately, and encrypt 
+footer by utilizing the `required i64 file_offset` parameter in the `ColumnChunk` 
+structure. This allows to serialize each `ColumnMetaData` structure separately and encrypt 
 it with a column-specific key, thus protecting the column stats and other metadata. 
 
+```c
+struct ColumnChunk {
+...
+  /** Byte offset in file_path to the ColumnMetaData **/
+  2: required i64 file_offset
+
+  /** Column metadata for this chunk. This is the same content as what is at
+   * file_path/file_offset.  Having it here has it replicated in the file
+   * metadata.
+   **/
+  3: optional ColumnMetaData meta_data
+..
+}
+```
+
 ### Encrypted footer mode
 
 In files with sensitive column data, a good security practice is to encrypt not only the 
-secret columns, but also the file footer metadata, with a separate footer key. This hides
+secret columns, but also the file footer metadata with a separate footer key. This hides
 the file schema / column names, number of rows, key-value properties, column sort order, 
 column data offset and size, list of encrypted columns and metadata of the column encryption keys. 
 It also makes the footer tamper-proof.
 
-The columns encrypted with the same key as the footer, don't split the ColumnMetaData from the 
+The columns encrypted with the same key as the footer don't separate the ColumnMetaData from the 
 ColumnChunks, leaving it at the original location, `optional ColumnMetaData meta_data`. This field
-is not set for columns enrypted with a column-specific key.
+is not set for columns encrypted with a column-specific key.
 
 A Thrift-serialized `FileCryptoMetaData` structure is written after the footer. It contains 
 information on the file encryption algorithm and on the footer (offset in 
 the file, and an optional key metadata, with a maximal length of 512). Then 
-the length of this structure is written, as a 4-byte little endian integer. Then a final 
+the length of this structure is written as a 4-byte little endian integer, followed by a final 
 magic string, "PARE".
-
-Only the `FileCryptoMetaData` is written as a plaintext, all other file parts are protected
+The same magic bytes are written at the beginning of the file (offset 0).
+Parquet readers start file parsing by reading and checking the magic string. Therefore, the encrypted footer mode uses a new magic string ("PARE") in order to immediately inform legacy readers (expecting ‘PAR1’ bytes) that they can’t parse this file.
```

Comment:
I would say the purpose of a different magic string is not to inform legacy reader that they can't read the file (they would fail anyway), but instead to inform new readers to look for a file crypto metadata at the end of the file instead of a footer.

Reply:
Agreed, this is the main reason. Still, having an early and clear exception in legacy readers, informing of a wrong file type (instead of a more obscure exception later on) is helpful too. I'll update the text accordingly.